libraries/AP_HAL_ChibiOS/AnalogIn.cpp

```diff
@@ -206,233 +236,389 @@ void AnalogSource::_add_value(float v, float vcc5V)
+}
+/*
+  return scaling
+ */
+float AnalogIn::get_pin_scaling(uint8_t adc_index, uint8_t pin_index)
+{
+    switch(adc_index) {
+    case 0:
+        return pin_config[pin_index].scaling;
+#if defined(HAL_ANALOG2_PINS)
+    case 1:
+        return pin_config_2[pin_index].scaling;
+#endif
+#if defined(HAL_ANALOG3_PINS)
+    case 2:
+        return pin_config_3[pin_index].scaling;
+#endif
+    };
+    osalDbgAssert(false, "invalid adc_index");
+    return 0;
+}
+
 /*
   callback from ADC driver when sample buffer is filled
  */
+#if HAL_WITH_MCU_MONITORING
+static uint16_t min_vrefint, max_vrefint;
+#endif
 void AnalogIn::adccallback(ADCDriver *adcp)
 {
-    const adcsample_t *buffer = samples;
-
-    stm32_cacheBufferInvalidate(buffer, sizeof(adcsample_t)*ADC_DMA_BUF_DEPTH*ADC_GRP1_NUM_CHANNELS);
+    uint8_t index = get_adc_index(adcp);
+    if (index >= HAL_NUM_ANALOG_INPUTS) {
+        return;
+    }
+    const uint16_t *buffer = (uint16_t*)samples[index];
+    uint8_t num_grp_channels = get_num_grp_channels(index);
+#if STM32_ADC_DUAL_MODE
+    if (index == 0) {
+        stm32_cacheBufferInvalidate(buffer, sizeof(uint16_t)*ADC_DMA_BUF_DEPTH*num_grp_channels*2);
+    } else
+#endif
+    {
+        stm32_cacheBufferInvalidate(buffer, sizeof(uint16_t)*ADC_DMA_BUF_DEPTH*num_grp_channels);
+    }
     for (uint8_t i = 0; i < ADC_DMA_BUF_DEPTH; i++) {
-        for (uint8_t j = 0; j < ADC_GRP1_NUM_CHANNELS; j++) {
-            sample_sum[j] += *buffer++;
+        for (uint8_t j = 0; j < num_grp_channels; j++) {
+            sample_sum[index][j] += *buffer;
+
+#if HAL_WITH_MCU_MONITORING
+            if (j == (num_grp_channels-1) && index == 2) {
+                // record min/max for MCU Vcc
+                if (min_vrefint == 0 ||
+                    min_vrefint > *buffer) {
+                    min_vrefint = *buffer;
+                }
+                if (max_vrefint == 0 ||
+                    max_vrefint < *buffer) {
+                    max_vrefint = *buffer;
+                }
+            }
+#endif
+            buffer++;
+#if STM32_ADC_DUAL_MODE
+            if (index == 0) {
+                // also sum the second ADC for dual mode
+                sample_sum[1][j] += *buffer;
+                buffer++;
+            } 
+#endif
         }
     }
-    sample_count += ADC_DMA_BUF_DEPTH;
+#if STM32_ADC_DUAL_MODE
+    if (index == 0) {
+        // also set the sample count for the second ADC for dual mode
+        sample_count[1] += ADC_DMA_BUF_DEPTH;
+    }
+#endif
+    sample_count[index] += ADC_DMA_BUF_DEPTH;
+}
+
+uint8_t AnalogIn::get_adc_index(ADCDriver* adcp)
+{
+    if (adcp == &ADCD1) {
+        return 0;
+    }
+#if defined(HAL_ANALOG2_PINS) && !STM32_ADC_DUAL_MODE
+    if (adcp == &ADCD2) {
+        return 1;
+    }
+#endif
+#if defined(HAL_ANALOG3_PINS)
+    if (adcp == &ADCD3) {
+        return 2;
+    }
+#endif
+    osalDbgAssert(false, "invalid ADC");
+    return 255;
+}
+
+uint8_t AnalogIn::get_num_grp_channels(uint8_t index)
```

Comment:
Probably too late now but I think the code would be much clearer / safer if you made index an enum since its only ever 0-2 and in the future its never going to be much more than this anyway